creator-node/src/fileManager.js

```diff
@@ -102,50 +104,98 @@ async function saveFileToIPFSFromFS (req, srcPath, fileType, sourceFile, transac
+  // and if so, create the directory first
+  // E.g if the path is /file_storage/QmABC/Qm123, we check if
+  // /QmABC/Qm123 contains more than 2 parts and if so, it's a directory
+  if (filePath.split('/').length > 2) {
+    const dir = path.dirname(expectedStoragePath)
+    if (dir) {
+      await mkdir(dir, { recursive: true })
+    }
+  }
+
   // If file not already stored, fetch from IPFS and store at storagePath.
-  const ipfs = req.app.get('ipfsAPI')
   let fileBuffer = null
-  req.logger.info(`Storing file at ${expectedStoragePath} for track multihash ${multihash}`)
 
   // If multihash already available on local INode, cat file from local ipfs node
-  req.logger.info(`checking if ${multihash} already available on local ipfs node`)
+  req.logger.debug(`checking if ${multihash} already available on local ipfs node`)
   try {
-    fileBuffer = await ipfs.cat(multihash)
-    req.logger.info(`Retrieved file for ${multihash} from local ipfs node`)
+    fileBuffer = await Utils.ipfsCat(multihash, req, 500)
+    req.logger.debug(`Retrieved file for ${multihash} from local ipfs node`)
+    // Write file to disk.
+    await writeFile(expectedStoragePath, fileBuffer)
+    req.logger.info(`wrote file to ${expectedStoragePath}, obtained via ipfs cat`)
   } catch (e) {
     req.logger.info(`Multihash ${multihash} is not available on local ipfs node`)
   }
 
   // If file not already available on local INode, fetch from IPFS.
   if (fileBuffer === null) {
-    req.logger.info(`Attempting to get ${multihash} from IPFS`)
+    req.logger.debug(`Attempting to get ${multihash} from IPFS`)
     let output
     try {
-      output = await ipfs.get(multihash)
+      output = await Utils.ipfsGet(multihash, req, 1000)
+      if (output.length !== 1) throw new Error(`provided multihash ${multihash} must map to 1 file`)
+      fileBuffer = output[0].content
+      req.logger.debug(`retrieved file for multihash ${multihash} from path ${output[0].path}`)
+      // Write file to disk.
+      await writeFile(expectedStoragePath, fileBuffer)
+      req.logger.info(`wrote file to ${expectedStoragePath}, obtained via ipfs get`)
     } catch (e) {
-      throw new Error(`Failed to retrieve file for multihash ${multihash} from IPFS`)
+      req.logger.info(`Failed to retrieve file for multihash ${multihash} from IPFS`)
     }
-    if (output.length !== 1) throw new Error('Audius track segment multihash must map to 1 file')
-    fileBuffer = output[0].content
-    req.logger.info(`retrieved file for multihash ${multihash} from path ${output[0].path}`)
   }
 
-  // Write file to disk.
-  const storagePath = path.join(req.app.get('storagePath'), multihash)
-  assert.strictEqual(storagePath, expectedStoragePath)
-  req.logger.info(`writing file to ${storagePath}...`)
-  await writeFile(storagePath, fileBuffer)
-  req.logger.info(`wrote file to ${storagePath}`)
+  // if file is still null, try to fetch from other cnode gateways with onlyFS=true
+  if (fileBuffer === null) {
```

Comment:
I thought we were going to rely on FS first before using IPFS to fetch the file?

Reply:
that's for the gateway. the fetch logic for nodesync is a different flow